### PR TITLE
Implemented CLI upload functionality

### PR DESCRIPTION
**Pull request by martinbrose**
Addressed #1543:

- #1543

### CODE REVIEW

File: docs/source/en/guides/upload.md [hunk 1]
```diff
@@ -106,6 +106,59 @@ but before that, all previous logs on the repo on deleted. All of this in a sing
 ... )
 ```
 
+## Upload from the CLI
+
+You can also upload files to the Hub directly from your terminal using the `huggingface-cli upload` command. Internally
```
Comment (stevhliu) [inline docs/source/en/guides/upload.md:111]:
```suggestion
You can use the `huggingface-cli upload` command from the terminal to directly upload files to the Hub. Internally
```

File: docs/source/en/guides/upload.md [hunk 2]
```diff
@@ -106,6 +106,59 @@ but before that, all previous logs on the repo on deleted. All of this in a sing
...
+# Usage:  huggingface-cli upload [repo_id] [local_path] [path_in_repo]
+>>> huggingface-cli upload Wauplin/my-cool-model ./models/model.safetensors model.safetensors
+https://huggingface.co/Wauplin/my-cool-model/blob/main/model.safetensors
+
+>>> huggingface-cli upload Wauplin/my-cool-model ./models .
+https://huggingface.co/Wauplin/my-cool-model/tree/main
+```
+
+`local_path` and `path_in_repo` are optional and can be implicitly inferred. By default, `local_path` will be set to
+the current directory and `path_in_repo` will be set to the relative path between the current directory and `local_path`.
```
Comment (stevhliu) [inline docs/source/en/guides/upload.md:126]:
```suggestion
`local_path` and `path_in_repo` are optional and can be implicitly inferred. By default, `local_path` is set to
the current directory, and `path_in_repo` is set to the relative path between the current directory and `local_path`.
```

**Review by stevhliu — APPROVED** [review]
Super cool to be able to schedule your upload with this as well! 👍

**Review by LysandreJik — APPROVED** [review]
The implementation looks great to me! 

I've added some comments regarding the API coming from linux expectations, but feel free to disregard if you feel strongly. The implementation is clean and it'll greatly simplify model uploading outside of python runtimes.

Thanks both for your work!

File: docs/source/en/guides/upload.md [hunk 3]
```diff
@@ -106,6 +106,59 @@ but before that, all previous logs on the repo on deleted. All of this in a sing
...
+If the implicit paths cannot be inferred, an error is raised.
+
+```bash
+# Upload file (implicit path_in_repo)
+huggingface-cli upload my-cool-model model.safetensors
+
+# Upload directory (implicit path_in_repo)
+huggingface-cli upload my-cool-model ./models
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/guides/upload.md]: Updated the CLI and now that's the case. Also added a unit test for this use case.

File: docs/source/en/guides/upload.md [hunk 4]
```diff
@@ -106,6 +106,59 @@ but before that, all previous logs on the repo on deleted. All of this in a sing
...
+
+# Upload directory (implicit local_path, implicit path_in_repo)
+huggingface-cli upload my-cool-model
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply docs/source/en/guides/upload.md:138]: Thanks for the feedback @LysandreJik. My initial thought was that `huggingface-cli upload my-cool-model` would upload the current directory to the Hub (at root level). So `huggingface-cli upload my-cool-model` being equivalent to `huggingface-cli upload my-cool-model . .` . A bit as when you have a local repo and you do `git add . && git commit -m "something" && git push`. 

That being said, the use case you've just described (e.g. repo_id == name of a local folder) is perfectly valid as well. I've updated the CLI in that sense. Now the behavior looks like this:

```bash
# Current repo structure
# ./
# ../
# folder-1/
#     file.txt
# folder-2/
#     file.txt

# Upload "./folder-1" content to "Wauplin/folder-1"
# On the hub: file.txt
>>> huggingface-cli upload folder-1

# Upload "./folder-1" content to "huggingface/folder-1"
# On the hub: file.txt
>>> huggingface-cli upload huggingface/folder-1

# Upload "./folder-1" and "./folder-2" to "Wauplin/my-cool-model"
# On the hub: folder-1/file.txt and folder-2/file.txt
>>> huggingface-cli upload my-cool-model .

# Upload "./folder-1" and "./folder-2" to "Wauplin/my-cool-model" under "./data"
# On the hub: data/folder-1/file.txt and data/folder-2/file.txt
>>> huggingface-cli upload my-cool-model . data/

# Raise exception => user must set local path explicitly 
>>> huggingface-cli upload folder-3
```

**Comment by Wauplin** [comment]
Thanks for the great feedback @LysandreJik! I have revisited a bit the API given your comments:
- No more relative path to compute the path_in_repo. Files are always uploaded at root level except if explicitly set differently by the user. It makes things easier to explain/understand as well.
- If the `local_path` is not provided (e.g. `huggingface-cli upload my-cool-model`), we check if `./my-cool-model` is a local file or folder. If it's the case, we upload it. If not, we raise an exception and ask the user to be explicit.

TL;DR: no implicit stuff except when it's _obvious_

I'll merge this as soon as the CI is green :)

**Comment by LysandreJik** [comment]
Thank you both for your work! :raised_hands:
